nrf_rpc/include/nrf_rpc.h

```diff
@@ -0,0 +1,444 @@
+/*
+ * Copyright (c) 2020 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+
+#ifndef _NRF_RPC_H_
+#define _NRF_RPC_H_
+
+#include <stdint.h>
+#include <stdbool.h>
+#include <stddef.h>
+
+#include <nrf_rpc_errors.h>
+#include <nrf_rpc_common.h>
+#include <nrf_rpc_tr.h>
+
+
+/**
+ * @defgroup nrf_rpc nRF RPC (Remote Procedure Calls) module.
+ * @{
+ * @brief Module to call procedures on a remote processor.
+ */
+
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+
+/** @brief First id of the nRF RPC group that can be used by the user */
+#define NRF_RPC_USER_GROUP_ID_FIRST 64
+
+
+/* Internal definition used in macros */
+#define _NRF_RPC_HEADER_SIZE 2
+
+
+/** @brief Callback that handles decoding of commands, events and responses.
+ *
+ * @param packet       Packet data.
+ * @param len          Length of the packet.
+ * @param handler_data Custom handler data. In case of commands, events it is a
+ *                     pointer to @a nrf_rpc_decoder structure associated with
+ *                     this command or event. In case of response handler it is
+ *                     an opaque pointer provided to @a NRF_RPC_CMD_SEND.
+ */
+typedef int (*nrf_rpc_handler_t)(const uint8_t *packet, size_t len,
+	void *handler_data);
+
+
+/** @brief Command and event decoder structure.
+ *
+ * Created by @a NRF_RPC_CMD_DECODER or @a NRF_RPC_EVT_DECODER.
+ */
+struct nrf_rpc_decoder {
+
+	/** @brief Command or event id. */
+	uint8_t id;
+
+	/** @brief Command or event decoder. */
+	nrf_rpc_handler_t handler;
+
+	/** @brief Command or event data for decoder. */
+	void *handler_data;
+};
+
+
+/** @brief Defines a group of commands and events.
+ *
+ * One group is mostly assigned to a single API that needs to be serialized.
+ * Created by @a NRF_RPC_GROUP_DEFINE.
+ */
+struct nrf_rpc_group {
+	uint8_t group_id;
+	const void *cmd_array;
+	const void *evt_array;
+};
+
+
+/** @brief Contains information about remote endpoint.
+ */
+struct nrf_rpc_remote_ep {
+	struct nrf_rpc_tr_remote_ep tr_ep;
+	uint8_t current_group_id;
+};
+
+
+/** @brief Contains information about local endpoint.
+ */
+struct nrf_rpc_local_ep {
+	struct nrf_rpc_tr_local_ep tr_ep;
+	struct nrf_rpc_remote_ep *default_dst;
+	uint32_t cmd_nesting_counter;
+	nrf_rpc_handler_t handler;
+	void *handler_data;
+};
+
+
+/** @brief Type of context variable for allocated resources.
+ *
+ * Variable of this type should be used as a context for @a NRF_RPC_CMD_ALLOC,
+ * @a NRF_RPC_EVT_ALLOC and @a NRF_RPC_RPS_ALLOC. After that should be passed
+ * to any of the send functions.
+ */
+typedef struct nrf_rpc_tr_remote_ep *nrf_rpc_alloc_ctx;
+
+
+/** @brief Define a group of commands and events.
+ *
+ * @param _name  Symbol name for the group.
+ * @param _id    Unique identified of the group. Can be from 0 to 127.
```

Comment:
Yes, I will add it.

Reply:
It was previously suggested that this kind of id may be a string. This will reduce need to maintain unique integer ids. With current implementation it is easy to do it, so I will send a new implementation of this macro in the next commit for this PR.